java/core/src/main/java/com/microsoft/bond/SpecificTypeObjectBonded.java

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+package com.microsoft.bond;
+
+import com.microsoft.bond.helpers.ArgumentHelper;
+import com.microsoft.bond.protocol.ProtocolWriter;
+
+import java.io.IOException;
+
+/**
+ * An implementation of {@link Bonded} backed by an object of a specific type that is known
+ * when the Bonded instance is constructed.
+ *
+ * @param <T> the specific struct type of this Bonded instance
+ */
+final class SpecificTypeObjectBonded<T extends BondSerializable> extends Bonded<T> {
+
+    private final T objectInstance;
+    private final StructBondType<T> bondType;
+
+    SpecificTypeObjectBonded(T objectInstance, StructBondType<T> bondType) {
+        this.objectInstance = objectInstance;
+        this.bondType = bondType;
+    }
+
+    @Override
+    public final StructBondType<T> getBondType() {
+        return this.bondType;
+    }
+
+    @Override
+    public void serialize(ProtocolWriter protocolWriter) throws IOException {
+        ArgumentHelper.ensureNotNull(protocolWriter, "protocolWriter");
+        this.bondType.serialize(this.objectInstance, protocolWriter);
+    }
+
+    @Override
+    public <U extends BondSerializable>
+    void serialize(ProtocolWriter protocolWriter, StructBondType<U> asBondType) throws IOException {
+        ArgumentHelper.ensureNotNull(protocolWriter, "protocolWriter");
+        ArgumentHelper.ensureNotNull(asBondType, "asBondType");
+        if (this.bondType.isSubtypeOf(asBondType)) {
+            // we can always serialize the base struct type portion by slicing
+            @SuppressWarnings("unchecked")
+            U castObjectInstance = (U) this.objectInstance;
+            asBondType.serialize(castObjectInstance, protocolWriter);
+        } else {
+            // but we need to transcode in all other cases
+            // TODO: complete this implementation after transcoding is implemented
+            throw new UnsupportedOperationException(
+                    "Serialization as a non-ancestor struct type is not currently implemented.");
```

Comment:
Please refer to .NET implementation for details. This case is supported in other Bond versions and thus needs to be supported in Java. Please note that this is only a limited Bonded implementation, as was necessary for tests and initial release and we will certainly reiterate on this in near future.

---
In reply to: [130510028](https://github.com/Microsoft/bond/pull/540#discussion_r130510028) [](ancestors = 130510028)